vfx/gl/texture_program.h

```diff
@@ -0,0 +1,73 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef VFX_GL_TEXTURE_PROGRAM_H_
+#define VFX_GL_TEXTURE_PROGRAM_H_
+
+#include "base/macros.h"
+#include "ui/gfx/geometry/point_f.h"
+#include "ui/gl/gl_bindings.h"
+#include "vfx/geometry/color.h"
+#include "vfx/geometry/matrix.h"
+#include "vfx/geometry/point.h"
+#include "vfx/gl/program.h"
+#include "vfx/gl/shader.h"
+
+namespace vfx {
+
+class TextureProgram {
+ public:
+  TextureProgram();
+  ~TextureProgram();
+
+  struct Vertex {
+    Point point;
+    Color color;
+    gfx::PointF tex_coord;
+  };
+
+  GLuint id() const { return program_.id(); }
+
+  template<typename Buffer>
+  void Draw(const Matrix& transform,
+            GLuint texture_id,
+            const Buffer& geometry) {
+    const GLvoid* kPositionOffset = nullptr;
+    const GLvoid* kColorOffset = reinterpret_cast<GLvoid*>(sizeof(GLfloat) * 3);
+    const GLvoid* kTexCoordOffset =
+        reinterpret_cast<GLvoid*>(sizeof(GLfloat) * 7);
+
+    glActiveTexture(GL_TEXTURE0);
+    glBindTexture(GL_TEXTURE_2D, texture_id);
+
+    glUseProgram(program_.id());
+    glUniformMatrix4fv(u_transform_, 1, GL_FALSE, transform.data());
+    glUniform1i(u_texture_, 0);
+    geometry.Bind();
+    glVertexAttribPointer(a_position_, 3, GL_FLOAT, GL_FALSE, sizeof(Vertex),
```

Comment:
So in your shader a_position is a vec4, but here you tell it that its 3 GL_FLOAT's, so the last component is gonna be undefined. If you're trying to pass in vec3's and add the 4th component in the vertex shader you should declare the type of a_position as vec3 in the shader and say 

`uniform mat4 u_transform;`
`attribute vec3 a_position;`
`...`
`gl_Position = u_transform * vec4(a_position, 1.0); `